### PR TITLE
Exercise 4 - Setup a CI Pipeline.

### PR DESCRIPTION
**Pull request by aashirathore6362**
Implement the CI and CD using GitHub Actions workflow.

### CODE REVIEW

File: .github/workflows/std-workflows.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+name: GitHub Actions CI/CD
+run-name: Deploying Local CI.
+on:
+  push:
+    branches:
+      - exercise-04
```
Comment (sitaramshelke) [inline .github/workflows/std-workflows.yml:6]: I think should be changed to match all the branches. 
Remember, unless the action is present in the branch, it won't be detected by it.

File: .github/workflows/std-workflows.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_ACCESS_TOKEN }}
+
+      - name: Building flask application.
+        run: |
+          make flaskapp_build
+
+      - name: Run docker compose.
```
Comment (sitaramshelke) [inline .github/workflows/std-workflows.yml:26]: This should be named docker build

File: .github/workflows/std-workflows.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        run: |
+          make docker_build
+
+      - name: Push the Image into dockerHub.
+        run: |
+          make docker_push
+        env:
+          REPO_NAME: aashirathore111
```
Comment (sitaramshelke) [inline .github/workflows/std-workflows.yml:34]: your understanding of Docker REPOSITORY is incorrect.

Please learn about the difference between a Docker Registry, Docker Username and Docker Repository.

File: makefile [hunk 4]
```diff
@@ -1,8 +1,21 @@
...
-	docker run -it -e FLASK_RUN_HOST=$(HOST) -e FLASK_RUN_PORT=$(PORT) -p $(PORT):$(PORT) studentapi:$(IMAGE_TAG)
+
+run:install-dep
+
+	flask -A student.py run --host=${HOSTPORT} --port=${APP_PORT}
+run-app:run
+
+#build application docker container.
+flaskapp_build:
+			docker build -t studentapi:1.0.0 .
```
Comment (sitaramshelke) [inline makefile:11]: Why is Image tag hardcoded? This way you will never have a different tag for your image

**Review by sitaramshelke — CHANGES_REQUESTED** [review]

File: .github/workflows/std-workflows.yml (hunk 1 above)
Comment (aashirathore6362) [reply .github/workflows/std-workflows.yml:6]: It's mean this should me trigger all the branches ?

File: makefile (hunk 4 above)
Comment (aashirathore6362) [reply makefile]: For this, I created env in github for passing image tag to makefile.

File: .github/workflows/std-workflows.yml (hunk 3 above)
Comment (aashirathore6362) [reply .github/workflows/std-workflows.yml]: I read some of the points I mentioned in my Trello card for this.
To resolve this I pass the username parameter to makefile.

File: .github/workflows/std-workflows.yml (hunk 1 above)
Comment (sitaramshelke) [reply .github/workflows/std-workflows.yml:6]: yes, The goal is to run CI for all branches? unless you have any other strategy

File: makefile [hunk 5]
```diff
@@ -1,7 +1,16 @@
...
+
+run:install-dep
+	flask -A student.py run --host=${HOSTPORT} --port=${APP_PORT}
+
+run-app:run
+
+#Push the docker image
+docker_push:
+	docker tag studentapi:$(IMAGE_TAG) ${USERNAME}/studentapi:1.0.0
+	docker push ${USERNAME}/studentapi:1.0.0
```
Comment (sitaramshelke) [inline makefile:12]: This should be `docker push ${USERNAME}/studentapi:${IMAGE_TAG}`

**Review by sitaramshelke — CHANGES_REQUESTED** [review]

File: makefile (hunk 5 above)
Comment (aashirathore6362) [reply makefile:12]: Tagging a image is required to push into docker hub.

**Pull request merged**